design/proposals/modular-components.md

```diff
@@ -0,0 +1,163 @@
+   that need to be tweaked across environments (_i.e._, _prod_ accessing a
+   different _mysql_ server than _dev_) 
+3. Configurations that are similar, however, have large pieces of code that
+   differ between environments and are not easily customizable by parameters.
+4. Configurations that are distinctively used in a single environment.
+
+ksonnet currently satisfies the use cases _1_ and _2_, however fall short on
+_3_ and _4_.
+
+## Goal
+
+* Redundancy between component manifests should be kept minimal.
+* It should be simple for an environment to only use a subset of all component
+  manifests.
+* It should be trivial to _add_ or _remove_ components from an environment.
+* It should be trivial to _clone_ an environment.
+* Environments should store the minimal amount of data necessary to operate
+  a set of _components_ on a _k8s cluster_.
+
+## Proposal
+
+1. Retire `main.jsonnet` and `base.libsonnet`.
+   The purpose of these files are often confusing to users. The goal of 
+   `main.jsonnet` was to act as a way for environments to provide
+   environment-specific extensibility to components. However, this only works
+   if the component is written in jsonnet, and also requires the user to have
+   an advanced understanding of jsonnet. Furthermore, with the existing model,
+   all components are operated on during `apply`, `delete`, etc. If the
+   components vary drastically between environments, it is not convenient to
+   "subtract" these components from each environment's `main.jsonnet`.
+   `base.libsonnet` also currently serves no purpose besides acting as an entry
+   point that _imports_ all components.
+2. Move environment metadata.
+   Each environment's `.metadata` directory currently store 3 files related to
+   the generated `ksonnet-lib`. This is unecessary and also costly as the
+   number of environments grow. We needn't store multiple copies of the same
+   API version on disk. The Kubernetes API version that each environment uses
+   can be recorded in the environment specification file. The metadata files
+   can be locally cached once per k8s API version in `lib`.
+3. A component should not be uniquely defined by it's base file name.
+   [More details soon]
+4. Handle environment specifications in `app.yaml` and retire `spec.json`.
+   Multiple files and areas for configuration can cause confusions for users.
+   In order to keep configurations simple, I propose that all ksonnet-specific
+   _app-level_ configurations reside in the `app.yaml`.
+
+`app.yaml` would add an _environments_ section as follows:
+
+```
+apiVersion: 0.0.1
+kind: ksonnet.io/app
+name: config
+version: 0.0.1
+registries:
+  incubator:
+    gitVersion:
+      commitSha: ea3408d44c2d8ea4d321364e5533d5c60e74bce0
+      refSpec: master
+    protocol: github
+    uri: github.com/ksonnet/parts/tree/master/incubator
+environments:
+  dev:
+    k8sVersion: 1.9.0
+    server: https://my-dev-server.com
+    namespaces:
```

Comment:
I propose we use a single namespace until Kubernetes has a better concept of managing multiple resources belonging to a single app across namespaces. If you want to go to another namespace, create another environment.

Reply:
+1 to simplicity

Reply:
Yes, the idea is that multiple namespaces should only be specified for an environment if the anticipation is that all components / params across the namespaces are identical. The scenario where multiple resources should be split across namespaces, should be modeled as multiple environments.

Reply:
I think supporting many namespaces where everything is uniform is a pretty typical use case. I can imagine one such example being to setup common RBAC policies in every namespace, setting up an ingress controller per namespace, and doing quotas/limitranges. Depending on what's being done, it may not always be the same by namespace, but in some of these cases, it might be.